### DISCUSSION TITLE
TypeScript support directly inside script tags

### DISCUSSION THREAD

**Discussion by Princesseuh** [withastro/roadmap#193] [Proposal]
## Background

Astro is very strongly TypeScript by default. Both the frontmatter and the template are TypeScript by default with no way of opt-ing out. However, there's one part of Astro file that is not that way: Script tags

```astro
---
  // This work
  const hello: string = "Hello"
  console.log(hello)
---

<script>
  // This doesn't
  const hello: string = "Hello"
  console.log(hello)
</script>
```

The only way you can use TypeScript inside a script tag right now is by doing the following:

```astro
<script>
  import "./script.ts"
</script>
```

> Note [the need for the extension](https://github.com/withastro/rfcs/discussions/191) :eyes: 

In my opinion, it's a bit awkward that you can import a `.ts` file (something totally non-standard) but you can't directly write TypeScript at the same place :sweat_smile: 

Before implementing [script tags support inside the editor tooling](https://github.com/withastro/language-tools/pull/263), I actually always thought our script tags were TypeScript, it felt intuitive to me since the rest of the file is

![image](https://user-images.githubusercontent.com/3019731/167013197-62d535da-ccfc-4571-ad08-659ddc3a3f1a.png)

## Proposal

In an effort of to improve consistency throughout the file, TypeScript should either be supported out of the box inside script tags or users should at least be able to opt-in TypeScript support using an attribute (`is:typescript`, `lang="ts"`, etc) on the script tag 

This should only affect bundled, processed etc script tags. So script tags using `is:inline` or `is:raw` would stay normal JavaScript ones. 

### Editor integration

Editor integration would follow suit, JavaScript script tags would only be checked if `checkJS` is set to `true` (or a ts comment is used) whereas TypeScript ones would be checked automatically like the frontmatter and the template without needing any user configuration

## Links

- [Vue's `lang="ts"`](https://vuejs.org/guide/typescript/overview.html#usage-in-single-file-components)
- [Svelte's `lang="ts"`](https://svelte.dev/blog/svelte-and-typescript#what-does-it-mean-to-support-typescript-in-svelte)
- [typescript-compile (`<script type="text/typescript">`)](https://github.com/niutech/typescript-compile)

**Comment by delucis** [comment]
Strongly in favour of out-of-the-box Typescript here!

**Comment by Princesseuh** [comment]
A new piece of information:

If you write this:
```astro
<script lang="ts">
  const hello: string = "Hello"
  console.log(hello)
</script>
```

It'll build, however the TS won't be transpiled. Could it be possible that this used to work and this is in fact a regression?

**Comment by Princesseuh** [comment]
We had an internal talk about this in the team and decided to support it, as it's deemed a bug and an inconsistency

This should ship as part of 1.0! :partying_face: :tada:
